src/renderer/base/renderer.cpp

```diff
@@ -526,6 +526,35 @@ void Renderer::TriggerFontChange(const int iDpi, const FontInfoDesired& FontInfo
     _NotifyPaintFrame();
 }
 
+// Routine Description:
+// - Called when the active soft font has been updated.
+// Arguments:
+// - bitPattern - An array of scanlines representing all the glyphs in the font.
+// - cellSize - The cell size for an individual glyph.
+// - centeringHint - The horizontal extent that glyphs are offset from center.
+// Return Value:
+// - <none>
+void Renderer::UpdateSoftFont(const gsl::span<const uint16_t> bitPattern, const SIZE cellSize, const size_t centeringHint)
+{
+    // We reserve PUA code points U+EF20 to U+EF7F for soft fonts, but the range
+    // that we test for in _IsSoftFontChar will depend on the size of the active
+    // bitPattern. If it's empty (i.e. no soft font is set), then nothing will
+    // match, and those code points will be treated the same as everything else.
+    const auto softFontCharCount = cellSize.cy ? bitPattern.size() / cellSize.cy : 0;
+    _lastSoftFontChar = _firstSoftFontChar + softFontCharCount - 1;
+
+    for (const auto pEngine : _rgpEngines)
+    {
+        LOG_IF_FAILED(pEngine->UpdateSoftFont(bitPattern, cellSize, centeringHint));
+    }
+    TriggerRedrawAll();
+}
+
+bool Renderer::_IsSoftFontChar(const std::wstring_view v) const

```

Comment:
This is going to sound like the rantings of a crazy person, but we learned that MSVC has some _unfortunate code gen_ when confronted with a string_view copy. References proved to be slightly better, and since we're doing this per character in the hot path it might be wise to take every win we can get.

@lhecker knows more, but I do not recall whether he's available right now.

Reply:
At the same time, perhaps this is trivially inlinable and doesn't contribute its fcall overhead?

Reply:
Hmmm... this looks like a no-win situation. The current code is inlined - in both places it's used - but the generated assembly is possibly not ideal. It looks to me like it's doing some unnecessary copying. If I change it to a reference, though, the generated assembly is better, but the second call (inside the inner loop) is no longer inlined. Maybe there's a way to get the best of both worlds, but it's not immediately obvious to me. I'll do a bit more experimenting, but you may need the experts for this.

Reply:
This may be a terrible idea...

```cpp
#define Renderer__Paamayim_Nekudotayim__IsSoftFontChar(v) (v.size() == 1 && v.front() >= _firstSoftFontChar && v.front() <= _lastSoftFontChar)
```

(yes, i have made myself sad)

Reply:
OK I've just tried that, and it does actually produce the best results. I just can't bring myself to use a define for this. Please let there be a better solution.

Reply:
I'm not sure if this classifies as better, but I can get the equivalent assembly that the `#define` produces by turning the `_IsSoftFontChar` method into a static, and passing in the `_firstSoftFontChar`  and `_lastSoftFontChar` fields as additional parameters. I don't like it, but I dislike it less than the `#define`.

Reply:
Just so everyone knows what kind of difference we're talking about here, this is a snippet of the assembly from the first part of the code (the rest is identical).

Original implementation:
```
00007FF6EF447638  movups      xmm0,xmmword ptr [r13]  
00007FF6EF44763D  movaps      xmmword ptr [rbp-80h],xmm0  
00007FF6EF447641  psrldq      xmm0,8  
00007FF6EF447646  movq        rax,xmm0  
00007FF6EF44764B  cmp         rax,1  
00007FF6EF44764F  jne         Microsoft::Console::Render::Renderer::_PaintBufferOutputHelper+36Fh (07FF6EF44766Fh)  
00007FF6EF447651  mov         rax,qword ptr [rbp-80h]  
00007FF6EF447655  movzx       ecx,word ptr [rax]  
00007FF6EF447658  cmp         rcx,qword ptr [r14+0B0h]  
```

Static function or #define:
```
00007FF69B127638  cmp         qword ptr [r13+8],1  
00007FF69B12763D  jne         Microsoft::Console::Render::Renderer::_PaintBufferOutputHelper+35Dh (07FF69B12765Dh)  
00007FF69B12763F  mov         rax,qword ptr [r13]  
00007FF69B127643  movzx       ecx,word ptr [rax]  
00007FF69B127646  cmp         rcx,qword ptr [r14+0B0h]  
```

Reply:
A couple more potential optimisations.

1. If we make the `_firstSoftFontChar` a `static constexpr`, the final memory comparison in the code above becomes a register comparison (albeit with an additional `mov`), which I'm assuming would be slightly more efficient.

    ```
    00007FF6DFAF7627  mov         eax,0EF20h  
    00007FF6DFAF762C  cmp         cx,ax  
    ```

2. If the conditions are reordered so the string size is compared last, we're more likely to fall through on the first test in typical usage (most characters would be less than `0xEF20`). However, this is assuming the size will always be at least 1, otherwise we could potentially be reading past the end of the string. I don't know enough about the buffer storage to know if that's possible.

Anyway, let me know if you want me to make any of these changes, or whether I should just leave it for the experts to optimise in a followup PR.

Reply:
Let's go with the middle ground version, which I've decided to be the `static` one. Then we can optimize from there :smile:

Reply:
Done.